cai_causal_graph/identify_utils.py

```diff
@@ -92,30 +122,138 @@ def _identify_confounders_no_checks_no_descendant_pruning(clean_graph: CausalGra
+    candidate_nodes = set(graph.get_ancestors(source_id)).difference(confounders)
+
+    # reject all candidate nodes that have a directed path to any confounders, and vice versa
+    for confounder in confounders:
+        confounder_descendants = graph.get_descendants(confounder)
+        confounder_ancestors = graph.get_ancestors(confounder)
+        for candidate in candidate_nodes.copy():
+            if candidate in confounder_descendants or candidate in confounder_ancestors:
+                candidate_nodes.remove(candidate)
+
+    # reject all candidate nodes that have a directed path to the destination that does not go through the source
+    for candidate in candidate_nodes.copy():
+        for causal_path in graph.get_all_causal_paths(candidate, destination_id):
+            if source_id not in causal_path:
+                candidate_nodes.remove(candidate)
+                break
+
+    # reject all candidate nodes that have a non-empty confounding set between itself and the destination
+    for candidate in candidate_nodes.copy():
+        if len(identify_confounders(graph, candidate, destination_id)) > 0:
+            candidate_nodes.remove(candidate)
+
+    return list(candidate_nodes)
+
+
+def identify_mediators(graph: CausalGraph, source: NodeLike, destination: NodeLike) -> List[str]:
+    """
+    Identify all mediators for the causal effect of `source` and `destination` in the provided `graph`.
+
+    A mediator variable for the causal effect of `source` on `destination` satisfies the following criteria:
+        1. There is a causal effect between the `source` and the `mediator`.
+        2. There is a causal effect between the `mediator` and the `destination`.
+        3. The `mediator` blocks all directed causal paths between the `source` and the `destination`.
+        4. There is no directed causal path from any confounder between `source` and `destination` to the `mediator`.
+
+    Example:
+        >>> from typing import List
+        >>> from cai_causal_graph import CausalGraph
+        >>> from cai_causal_graph.identify_utils import identify_mediators
+        >>>
+        >>> # define a causal graph
+        >>> cg = CausalGraph()
+        >>> cg.add_edge('x', 'm')
+        >>> cg.add_edge('m', 'y')
+        >>> cg.add_edge('u', 'x')
+        >>> cg.add_edge('u', 'y')
+        >>> cg.add_edge('x', 'y')
+        >>>
+        >>> # find the mediators between 'x' and 'y'; output: ['m']
+        >>> mediator_variables: List[str] = identify_mediators(cg, source='x', destination='y')
+
+    :param graph: The causal graph given by a `cai_causal_graph.causal_graph.CausalGraph` instance. This must be a DAG,
+        i.e. it must only contain directed edges and be acyclic, otherwise a `TypeError` is raised.
+    :param source: The source variable or its identifier.
+    :param destination: The destination variable or its identifier.
+    :return: A list of mediator variables for the causal effect of `source` on `destination`.
```

Comment:
Would this require the same caveat? e.g., when doing front door adjustment. 

Should we also add a caveat to return on confounders with example of back door adjustment? Say we are identifying all confounders but the total list may not be optimal one for later causal tasks, such as performing back door adjustment when estimating a causal effect.

Reply:
Yep, I agree, although the caveat here is that, really, they should be using all returned variables when adjusting for confounding, because it is a sufficient adjustment set. Intuitively, I would like to say that it is a minimal adjustment set, but I have no proof for that.

Reply:
This [talk](https://stat.ethz.ch/~maathuis/meetings/SlidesMaathuisNeurIPS2020.pdf) has some good stuff on optimal adjustment sets. The optimal is not always the maximal. But yeah agree for backdoor we need to block all backdoor paths so would need all confounders. For IV though don't need to use all.